File: GameOfLife.CSharp/Board.cs
using System.Collections.Generic;

namespace GameOfLife.CSharp
{
    public enum CellState
    {
        Dead,
        Alive,
    }

    public class Board
    {
        readonly CellState[] cells;
        public int Width { get; }
        public int Height { get; }

        public Board(int width, int height)
        {
            cells = new CellState[width * height];
            Width = width;
            Height = height;
        }

        public CellState this[int x, int y]
        {
            get => cells[y * Width + x];
            set => cells[y * Width + x] = value;
        }

        bool IsInRange(int x, int y) =>
            x >= 0 &&
            y >= 0 &&
            x < Width &&
            y < Height;

        public IEnumerable<CellState> GetNeighbors(int x, int y)
        {
            for (int i = x - 1; i <= x + 1; i++)
                for (int j = y - 1; j <= y + 1; j++)
                {
                    if (i == x && j == y)
                        continue;

                    if (IsInRange(i, j))
                        yield return this[i, j];
                }
        }
    }
}

File: GameOfLife.CppOpenMP/Board.cpp
#include "Board.h"

Board::Board(int width, int height) :
	width(width), height(height)
{
	cells = new CellState[width * height]{};
}

Board::~Board()
{
	delete cells;
}

int Board::GetWidth()
{
	return width;
}

int Board::GetHeight()
{
	return height;
}

int Board::GetLiveNeighbors(int x, int y)
{
	int liveNeighbors = 0;

	for (int i = x - 1; i <= x + 1; i++)
		for (int j = y - 1; j <= y + 1; j++)
		{
			if (i == x && j == y)
				continue;

			if (IsInRange(i, j) && Get(i, j) == Alive)
				liveNeighbors++;
		}

	return liveNeighbors;
}

CellState Board::Get(int x, int y)
{
	return cells[y * width + x];
}

Board& Board::Set(int x, int y, CellState state)
{
	cells[y * width + x] = state;
	return *this;
}

bool Board::IsInRange(int x, int y)
{
	return x >= 0 &&
		y >= 0 &&
		x < width &&
		y < height;
}
